### PR TITLE
Fixes: Bundle product with only one product in required bundle option not being added to cart with bundle option.

### LINKED ISSUE #38605: [Issue] Fixes: Bundle product with only one product in required bundle option not being added to cart with bundle option.
This issue is automatically created based on existing pull request: magento/magento2#38601: Fixes: Bundle product with only one product in required bundle option not being added to cart with bundle option.

---------
### Description (*)
Adding a bundle product to the cart from the product list widget doesn’t let the product be added properly with the bundle option. The bundle product options are not added within the parent product. It shows a warning message in minicart like ‘Please specify product option(s)’. It doesn’t fail in all the cases, it only fails in a certain case where a bundle product has been created with a required bundle option & the bundle option contains only one product with the ‘is_default’ checkbox selected.

Screenshot 1:
image
Here you can see in the screenshot a bundle product has been created with a bundle option containing only one product in it & the option is required and the product is default select.

Screenshot 2:
image
After creating a bundle product, a widget has been created with a type of Catalog Product List & a condition that matches the particular bundle product.

Screenshot 3:
image
As we have created a widget, a bundle product is visible on the home page in the widget.

Screenshot 4:
image
When we click on the add-to-cart button, the product seems to be added to the cart & it doesn’t throw any error. And then if you check in the mini-cart, you will come to see a warning message like “Please specify product option(s).” 

The same thing works fine on a list page, but it fails while adding a product to the cart through the widget.

### Manual testing scenarios (*)
1. Create a bundle product that contains one or more bundle options & each bundle option is required and must contain only one product that is default selected. (see screenshot 1)
2. Create a widget with the type ‘Catalog Product List’ with a condition that includes the newly created bundle product in the widget. (see screenshot 2)
3. From the front end visit the widget & try to add the bundle product from the widget product list.(see screenshot 3)
4. After clicking Add to Cart observe a warning message in minicart. (see screenshot 4).

### Expected result (*)
The bundle product should be added to the cart with bundle options.

### Actual result (*)
The bundle product isn’t being added to the cart with bundle options.

### Additional Information (*)
File1: `Magento/CatalogWidget/view/frontend/templates/product/widget/content/grid.phtml`
File2: `Magento/Catalog/view/frontend/templates/product/list.phtml`

If we check in the above-given list.phtml, it gets the option by using the ViewModel & which works fine on the list page. (see screenshot 5).
Screenshot 5:
image

In the given `grid.phtml` file it doesn’t get the options data where the add-to-cart form is rendered for the particular product (see screenshot 6). As the options are not rendered, adding that product to the cart will only add the parent bundle product to the cart without any bundle options. So, we need to implement the same thing as `list.phtml` to get the data in `grid.phtml`.

Screenshot 6:
image

### Solution (*)
We have to update the given phtml to get the bundle options in the add-to-cart form. We can get the bundle product options by using the `Magento\Catalog\ViewModel\Product\OptionsData` ViewModel. (see  screenshot 7)

A separate function has been created in the block file of the `grid.phtml` (`Magento\CatalogWidget\Block\Product\ProductsList`) to get the ViewModel that returns the object of `Magento\Catalog\ViewModel\Product\OptionsData`.

Screenshot 7:
image

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by bhoopatparmar**
### Description (*)
Adding a bundle product to the cart from the product list widget doesn’t let the product be added properly with the bundle option. The bundle product options are not added within the parent product. It shows a warning message in minicart like ‘Please specify product option(s)’. It doesn’t fail in all the cases, it only fails in a certain case where a bundle product has been created with a required bundle option & the bundle option contains only one product with the ‘is_default’ checkbox selected.

Screenshot 1:
image
Here you can see in the screenshot a bundle product has been created with a bundle option containing only one product in it & the option is required and the product is default select.

Screenshot 2:
image
After creating a bundle product, a widget has been created with a type of Catalog Product List & a condition that matches the particular bundle product.

Screenshot 3:
image
As we have created a widget, a bundle product is visible on the home page in the widget.

Screenshot 4:
image
When we click on the add-to-cart button, the product seems to be added to the cart & it doesn’t throw any error. And then if you check in the mini-cart, you will come to see a warning message like “Please specify product option(s).” 

The same thing works fine on a list page, but it fails while adding a product to the cart through the widget.

### Manual testing scenarios (*)
1. Create a bundle product that contains one or more bundle options & each bundle option is required and must contain only one product that is default selected. (see screenshot 1)
2. Create a widget with the type ‘Catalog Product List’ with a condition that includes the newly created bundle product in the widget. (see screenshot 2)
3. From the front end visit the widget & try to add the bundle product from the widget product list.(see screenshot 3)
4. After clicking Add to Cart observe a warning message in minicart. (see screenshot 4).

### Expected result (*)
The bundle product should be added to the cart with bundle options.

### Actual result (*)
The bundle product isn’t being added to the cart with bundle options.

### Additional Information (*)
File1: `Magento/CatalogWidget/view/frontend/templates/product/widget/content/grid.phtml`
File2: `Magento/Catalog/view/frontend/templates/product/list.phtml`

If we check in the above-given list.phtml, it gets the option by using the ViewModel & which works fine on the list page. (see screenshot 5).
Screenshot 5:
image

In the given `grid.phtml` file it doesn’t get the options data where the add-to-cart form is rendered for the particular product (see screenshot 6). As the options are not rendered, adding that product to the cart will only add the parent bundle product to the cart without any bundle options. So, we need to implement the same thing as `list.phtml` to get the data in `grid.phtml`.

Screenshot 6:
image

### Solution (*)
We have to update the given phtml to get the bundle options in the add-to-cart form. We can get the bundle product options by using the `Magento\Catalog\ViewModel\Product\OptionsData` ViewModel. (see  screenshot 7)

A separate function has been created in the block file of the `grid.phtml` (`Magento\CatalogWidget\Block\Product\ProductsList`) to get the ViewModel that returns the object of `Magento\Catalog\ViewModel\Product\OptionsData`.

Screenshot 7:
image

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#38605: Fixes: Bundle product with only one product in required bundle option not being added to cart with bundle option.

### Related Pull Requests
https://github.com/magento-gl/magento2-page-builder/pull/84

### CODE REVIEW

**Comment by bhoopatparmar** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento create issue

File: app/code/Magento/CatalogWidget/view/frontend/templates/product/widget/content/grid.phtml [hunk 1]
```diff
@@ -74,6 +74,12 @@ use Magento\Wishlist\Helper\Data;
                                                     <?php if ($_item->isSaleable()): ?>
                                                         <?php $postParams = $block->getAddToCartPostParams($_item); ?>
                                                         <form data-role="tocart-form" data-product-sku="<?= $escaper->escapeHtml($_item->getSku()) ?>" action="<?= $escaper->escapeUrl($postParams['action']) ?>" method="post">
+                                                            <?php $options = $block->getViewModel()->getOptionsData($_item); ?>
```
Comment (Bashev) [inline app/code/Magento/CatalogWidget/view/frontend/templates/product/widget/content/grid.phtml]: Will be complete solution if you add the same for the pagebuilder package (in their repository https://github.com/magento/magento2-page-builder) https://github.com/magento/magento2-page-builder/blob/4a2795ea9d2dae7cec8437e5e25e78e67c1b8d90/app/code/Magento/PageBuilder/view/frontend/templates/catalog/product/widget/content/carousel.phtml#L62

Comment (Bashev) [reply app/code/Magento/CatalogWidget/view/frontend/templates/product/widget/content/grid.phtml]: Same as here: https://github.com/magento/magento2-page-builder/pull/868/files#r1560503440
`<?php $options = $block->getViewModel()->getOptionsData($_item); ?>`
change it to 
`<?php if($options = $block->getViewModel()?->getOptionsData($_item)) : ?>`
will prevent empty/not assigned viewModel

and ofc, close if, at the end of foreach.

Comment (bhoopatparmar) [reply app/code/Magento/CatalogWidget/view/frontend/templates/product/widget/content/grid.phtml]: @Bashev Thanks for your suggestion. I've updated the code as per your suggestion

**Review by Bashev — APPROVED** [review]
Looks fine for me.

**Comment by engcom-Bravo** [comment]
@magento run all tests

**Comment by engcom-Bravo** [comment]
@magento give me test instance

**Review by engcom-Bravo — CHANGES_REQUESTED** [review]
Hi @bhoopatparmar,

Thanks for the collaboration & contribution!

:x: QA not Passed

**Manual Scenario Steps**

- Create a bundle product that contains one or more bundle options & each bundle option is required and must contain only one product that is default selected. (see screenshot 1)
- Create a widget with the type ‘Catalog Product List’ with a condition that includes the newly created bundle product in the widget. (see screenshot 2)
- From the front end visit the widget & try to add the bundle product from the widget product list.(see screenshot 3)
- After clicking Add to Cart observe a warning message in minicart. (see screenshot 4).

**:heavy_check_mark: Expected result After Fix** 

The bundle product should be added to the cart with bundle options

**:x: Actual result After Fix**



Screenshot 2024-04-15 at 16 50 16



Screenshot 2024-04-15 at 16 50 46


Bundle Products are still not adding to the cart with the Bundle options.

Could you please let us know if we are missing anything.

Thanks.

**Comment by bhoopatparmar** [comment]
Hi @engcom-Bravo 
getViewModel() method remained to add to the `Magento\CatalogWidget\Block\Product\ProductsList`, which led it to not getting the ViewModel object & options were not rendered. So, I've updated the code.

**Comment by bhoopatparmar** [comment]
@magento run all test

**Comment by bhoopatparmar** [comment]
@magento run all test

**Comment by nicholasfishjgm** [comment]
Any update on this please?

**Comment by bhoopatparmar** [comment]
@magento run all test

**Comment by Bashev** [comment]
@magento run all tests

**Comment by bhoopatparmar** [comment]
@magento run Static Tests

**Comment by bhoopatparmar** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: app/code/Magento/CatalogWidget/Block/Product/ProductsList.php [hunk 2]
```diff
@@ -15,6 +15,7 @@
 use Magento\Catalog\Model\ResourceModel\Product\Collection;
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogWidget/Block/Product/ProductsList.php:16]: In accordance with [Basic Programming Principles](https://developer.adobe.com/commerce/php/coding-standards/technical-guidelines/#1-basic-programming-principles), we should add `declare(strict_types=1);` for all update PHP files.

File: app/code/Magento/CatalogWidget/Block/Product/ProductsList.php [hunk 3]
```diff
@@ -83,6 +84,11 @@ class ProductsList extends AbstractProduct implements BlockInterface, IdentityIn
      */
     protected $productCollectionFactory;
 
+    /**
+     * @var OptionsData
+     */
+    protected $optionsDataViewModel;
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogWidget/Block/Product/ProductsList.php]: Please make this variable as `private`.

File: app/code/Magento/CatalogWidget/Block/Product/ProductsList.php [hunk 4]
```diff
@@ -162,6 +170,7 @@ public function __construct(
     ) {
         $this->productCollectionFactory = $productCollectionFactory;
         $this->catalogProductVisibility = $catalogProductVisibility;
+        $this->optionsDataViewModel = $optionsDataViewModel;
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogWidget/Block/Product/ProductsList.php]: In order to make this Backward compatible new params in Constructor should be null and handle with ObjectManager:

```suggestion
        $this->optionsDataViewModel = $optionsDataViewModel ?: ObjectManager::getInstance()->get(OptionsData::class);
```

File: app/code/Magento/CatalogWidget/Block/Product/ProductsList.php [hunk 5]
```diff
@@ -150,6 +157,7 @@ public function __construct(
         Context $context,
         CollectionFactory $productCollectionFactory,
         Visibility $catalogProductVisibility,
+        OptionsData $optionsDataViewModel,
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogWidget/Block/Product/ProductsList.php]: Make this null

```suggestion
        OptionsData $optionsDataViewModel = null,
```

File: app/code/Magento/CatalogWidget/Block/Product/ProductsList.php [hunk 6]
```diff
@@ -288,6 +297,16 @@ protected function getDetailsRendererList()
         return $this->rendererListBlock;
     }
 
+    /**
+     * Get the view model.
+     *
+     * @return OptionsData
+     */
+    public function getViewModel()
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogWidget/Block/Product/ProductsList.php]: Please define its return type.

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @bhoopatparmar,

Thanks for the contribution!

Please refer to the review comments below. Also please add some automated test in accordance to [DOD](https://developer.adobe.com/commerce/contributor/guides/code-contributions/definition-of-done/#definition-of-done-1).

Thanks

**Comment by bhoopatparmar** [comment]
Hi @engcom-Hotel 

Thanks for your suggestions. 
I've updated the code as per your suggestions & updated automated test.
Let me know if I misses anything.

thanks!

**Comment by bhoopatparmar** [comment]
@magento run all tests

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Thanks @bhoopatparmar, for the quick workaround.

But instead of the unit test, can we have a MFTF to check the functionality.

Thanks

**Comment by engcom-Echo** [comment]
@magento give me test instance

**Comment by engcom-Echo** [comment]
@bhoopatparmar,

Thanks for the collaboration & contribution!

:x: QA not Passed

**Manual Scenario Steps**

- Create a bundle product that contains one or more bundle options & each bundle option is required and must contain only one product that is default selected. (see screenshot 1)
- Create a widget with the type ‘Catalog Product List’ with a condition that includes the newly created bundle product in the widget. (see screenshot 2)
- From the front end visit the widget & try to add the bundle product from the widget product list.(see screenshot 3)
- After clicking Add to Cart observe a warning message in minicart. (see screenshot 4).

**:heavy_check_mark: Expected result After Fix** 

The bundle product should be added to the cart with bundle options

**:x: Actual result After Fix**
Screenshot 2024-06-04 at 4 03 59 PM


Bundle Products are still not adding to the cart with the Bundle options.


Thanks.

**Comment by engcom-Echo** [comment]
@bhoopatparmar as per the above [comment](https://github.com/magento/magento2/pull/38601#issuecomment-2147196994). we are moving this PR to `Draft` state.
Once you are done with changes you can move from `draft` status to `open`.

**Comment by bhoopatparmar** [comment]
@magento give me test instance

**Comment by bhoopatparmar** [comment]
@magento give me test instance

**Comment by bhoopatparmar** [comment]
@magento run all tests

**Comment by bhoopatparmar** [comment]
@magento give me test instance

**Comment by bhoopatparmar** [comment]
@magento give me test instance

**Comment by bhoopatparmar** [comment]
@magento give me test instance

**Comment by bhoopatparmar** [comment]
Hi @engcom-Echo 
Can you re-verify the solution ? I can see it is working as expected & I have tested it in the test instance as well.
see the video for reference 

https://github.com/magento/magento2/assets/106948523/652b6da5-2dc4-4994-b3b9-00927434cadd

Please provide a video for reference if you still faces the issue.
Thank you.

**Comment by engcom-Bravo** [comment]
@magento give me test instance

**Comment by engcom-Echo** [comment]
@magento give me test instance

**Comment by engcom-Echo** [comment]
@magento give me test instance

**Comment by engcom-Echo** [comment]
@magento give me test instance

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Review by engcom-Hotel — APPROVED** [review]

**Review by engcom-Bravo — CHANGES_REQUESTED** [review]
Hi @bhoopatparmar,

Thanks for Contribution!!.

Could you please resolve conflicts to proceed further with the PR.

Thanks.

**Comment by bhoopatparmar** [comment]
Hi @engcom-Bravo 
The solution has been merged in the dev branch by in this commit: https://github.com/magento/magento2/commit/d921385856b7fafb34238fec4724f4ea0570f082
Hence closing this PR
Thank you

**Pull request closed without merging**